servo-media/src/backends/gstreamer/audio_decoder.rs

```diff
@@ -0,0 +1,142 @@
+
+                match media_type {
+                    None => {
+                        eprintln!("Failed to get media type from pad {}", src_pad.get_name());
+                        // XXX reject Future.
+                        return;
+                    }
+                    Some(media_type) => media_type,
+                }
+            };
+
+            let insert_sink = || -> Result<(), ()> {
+                let queue = gst::ElementFactory::make("queue", None).ok_or(())?;
+                let convert = gst::ElementFactory::make("audioconvert", None).ok_or(())?;
+                let resample = gst::ElementFactory::make("audioresample", None).ok_or(())?;
+                // XXX Use audiosink for now. This should end up being an appsink.
+                let sink = gst::ElementFactory::make("autoaudiosink", None).ok_or(())?;
+
+                let elements = &[&queue, &convert, &resample, &sink];
+                pipeline.add_many(elements).map_err(|_| ())?;
+                gst::Element::link_many(elements).map_err(|_| ())?;
+
+                for e in elements {
+                    e.sync_state_with_parent().map_err(|_| ())?;
+                }
+
+                let sink_pad = queue.get_static_pad("sink").expect("queue has no sinkpad");
+                src_pad
+                    .link(&sink_pad)
+                    .into_result()
+                    .map(|_| ())
+                    .map_err(|_| ())
+            };
+
+            if !is_audio || insert_sink().is_err() {
+                // XXX reject Future.
+                return;
+            }
+        });
+
+        let appsrc = appsrc.downcast::<AppSrc>().map_err(|_| ())?;
+        appsrc.set_property_format(gst::Format::Time);
+        appsrc.set_property_block(true);
+
+        let _ = pipeline.set_state(gst::State::Playing);
+
+        // We push data into the appsrc in a different thread so we
+        // can get messages from the bus.
+
+        let appsrc_ = appsrc.clone();
+        Builder::new()
+            .name("Decoder data loop".to_owned())
+            .spawn(move || {
+                let mut offset = 0;
+                let max_bytes = appsrc_.get_max_bytes() as usize;
+                while offset < data.len() {
+                    let data_left = data.len() - offset;
+                    let buffer_size = if data_left < max_bytes {
+                        data_left
+                    } else {
+                        max_bytes
+                    };
+                    let mut buffer = gst::Buffer::with_size(buffer_size).unwrap();
+                    {
+                        let buffer = buffer.get_mut().unwrap();
+                        let next_offset = offset + buffer_size;
+                        let data = data.as_mut_slice();
+                        buffer
+                            .copy_from_slice(0, &data[offset..next_offset])
+                            .expect("copying failed");
+                        offset = next_offset;
```

Comment:
You could use `std::io::Cursor` and `Read` on the `Vec` to prevent having to keep track of the offset yourself